### PR TITLE
Mesh importer tangents with surface tool

### PR DESCRIPTION
**Pull request by fire**
Do not merge for 4.3.

https://github.com/godotengine/godot-proposals/issues/8696

- [x] Blend shapes have special processing

See https://github.com/godotengine/godot/pull/83648 for a test case.

### CODE REVIEW

File: servers/rendering_server.cpp [hunk 1]
```diff
@@ -1226,8 +1226,10 @@ Error RenderingServer::mesh_create_surface_data_from_arrays(SurfaceData *r_surfa
 					bsformat |= (1 << j);
 				}
 			}
-
-			ERR_FAIL_COND_V_MSG(bsformat != (format & RS::ARRAY_FORMAT_BLEND_SHAPE_MASK), ERR_INVALID_PARAMETER, "Blend shape format must match the main array format for Vertex, Normal and Tangent arrays.");
+			bool has_vertex = bsformat & RS::ARRAY_FORMAT_VERTEX;
+			bool has_normal = bsformat & RS::ARRAY_FORMAT_NORMAL;
+			bool has_tangent = bsformat & RS::ARRAY_FORMAT_TANGENT;
+			ERR_FAIL_COND_V_MSG(bsformat != (format & RS::ARRAY_FORMAT_BLEND_SHAPE_MASK), ERR_INVALID_PARAMETER, vformat("Blend shape format must match the main array format for Vertex, Normal and Tangent arrays. Vertex status: %s, Normal status: %s, Tangent status: %s, Current blend shape format: %d, Main array format: %d", has_vertex ? "OK" : "Not OK", has_normal ? "OK" : "Not OK", has_tangent ? "OK" : "Not OK", bsformat, format & RS::ARRAY_FORMAT_BLEND_SHAPE_MASK));
```
Comment (fire) [inline servers/rendering_server.cpp:1232]: Not essential code, can be move.

**Comment by clayjohn** [comment]
The linked PR appears to be for removing SurfaceTool from the import process, but this seems to increase reliance on the SurfaceTool instead of removing it. Is there a reason for the change in plan?

**Comment by fire** [comment]
The idea is we first consolidate all the uses of Surface Tool in one place in the scene importer from GLTF, Obj, FBX and Collada. Then, we can optimize that entire chunk of code as a whole by in-lining it etc. Like split the problem.

I read the proposal for this carefully, and although it said there is a desire to remove surface tool, it also said that duplication of code is an issue among the importers so there's two problems.

File: scene/resources/surface_tool.cpp [hunk 2]
```diff
@@ -1152,9 +1152,19 @@ void SurfaceTool::mikktSetTSpaceDefault(const SMikkTSpaceContext *pContext, cons
 }
 
 void SurfaceTool::generate_tangents() {
-	ERR_FAIL_COND(!(format & Mesh::ARRAY_FORMAT_TEX_UV));
 	ERR_FAIL_COND(!(format & Mesh::ARRAY_FORMAT_NORMAL));
-
+	if (!(format & Mesh::ARRAY_FORMAT_TEX_UV)) {
```
Comment (fire) [inline scene/resources/surface_tool.cpp:1156]: Is it a requirement to have tangents, some of the collada code generates tangents, but I have a desire to not generate them if wrong.

Comment (clayjohn) [reply scene/resources/surface_tool.cpp:1156]: Tangents are always required, if the mesh doesn't provide them then the renderer will generate them. Ideally tangent generation would happen at import time though

**Comment by fire** [comment]
## Conversation between Bqq and Lyuma on Tangents

1. **Broken Tangents and Blend Shape Normals in DCC:** Lyuma expressed concern about the perception that some Digital Content Creation (DCC) tools may produce broken tangents or blend shape normals+tangents.

2. **Godot's Import Settings Limitations:** According to Lyuma, Godot's import settings are too limited. Users have to choose between 'Ensure Tangents' or not. However, if blend shapes exist, this can completely break.

3. **bqq's Approach for Tangents:** bqq shared his personal approach for handling tangents, which is to discard the DCC tangents unconditionally and recompute new ones with mikktspace. He believes this should be an option, not the default.

4. **Godot's Lack of Option for Recomputing Tangents:** Lyuma pointed out that Godot doesn't provide the option to recompute tangents. If DCC tangents exist, Godot will always use them, leaving no way to fix these.

## Suggested Improvements
- A dropdown option for handling tangents like "always generate", "generate if missing", "ignore if missing", "remove always(?)".
- Improvement in Godot's import settings to handle tangents better.

File: editor/import/3d/resource_importer_scene.cpp [hunk 3]
```diff
@@ -2522,6 +2522,83 @@ Node *ResourceImporterScene::_generate_meshes(Node *p_node, const Dictionary &p_
...
+								Vector<Vector3> normals = arrays[Mesh::ARRAY_NORMAL];
+								Vector<float> tangents = arrays[Mesh::ARRAY_TANGENT];
+								for (int vert = 0; vert < normals.size(); vert++) {
+									Vector3 tan = Vector3(tangents[vert * 4 + 0], tangents[vert * 4 + 1], tangents[vert * 4 + 2]);
+									if (abs(tan.dot(normals[vert])) > 0.0001) {
+										// Tangent is not perpendicular to the normal, so we can't use compression.
+										format &= ~RS::ARRAY_FLAG_COMPRESS_ATTRIBUTES;
+									}
+								}
+							}
```
Comment (fire) [inline editor/import/3d/resource_importer_scene.cpp:2594]: In the previous code the tangent compression check is done outside.

**Comment by Calinou** [comment]
> A dropdown option for handling tangents like "always generate", "generate if missing", "ignore if missing", "remove always(?)".

A **Remove Always** option makes sense to minimize mesh file size/memory usage if you don't intend on using normal maps.

**Comment by lyuma** [comment]
> Remove Always option makes sense to minimize mesh file size/memory usage

The problem is that godot did something in 4.2 that made mesh format more rigid and broke many models unless they have normals and tangents. I don't want to add this until we fix the bugs in the rendering server

I understand the rendering team has a lot bigger issues to contend with right now but it would be nice to have a discussion with them on this issue at some point and figure out if there is a way to make mesh formats less rigid so we can optimize models that don't need tangents or normals.

**Comment by clayjohn** [comment]
> > Remove Always option makes sense to minimize mesh file size/memory usage
> 
> The problem is that godot did something in 4.2 that made mesh format more rigid and broke many models unless they have normals and tangents. I don't want to add this until we fix the bugs in the rendering server
> 
> I understand the rendering team has a lot bigger issues to contend with right now but it would be nice to have a discussion with them on this issue at some point and figure out if there is a way to make mesh formats less rigid so we can optimize models that don't need tangents or normals.

Can you be more specific what issues you are seeing? AFAIK we have fixed all the issues with tangents in the RenderingServer already. There aren't any more bug reports. If you are seeing issues arising from the new mesh compression stuff, I'd welcome a bug report. Fixing these tangent issues is a very high priority for me.

**Comment by clayjohn** [comment]
> > Can you be more specific what issues you are seeing? AFAIK we have fixed all the issues with tangents in the RenderingServer already. There aren't any more bug reports. If you are seeing issues arising from the new mesh compression stuff, I'd welcome a bug report. Fixing these tangent issues is a very high priority for me.
> 
> @clayjohn I am talking about the regression in 4.2 on meshes with blend shapes without tangents. That regression is still present in my 1-week-old build.
> Ensure Tangents on: works
> vrm_ensure_tangents_on
> 
> Ensure Tangents off: mesh fails to render
> vrm_ensure_tangents_off
> 
> I'm reluctant to offer an option in the import settings which will break meshes (I know we have Ensure Tangents, but we need to keep it for compatibility and it defaults to on)

Very interesting. Is there a bug report I can take a look at? I can fix this if I have access to a bug report or MRP. I haven't seen this issue reported before though

**Comment by lyuma** [comment]
@clayjohn documented in #86384
mrp: https://github.com/godotengine/godot/files/13738417/MeshTangents.zip

and mrp from 85960: https://github.com/godotengine/godot/files/13623984/resource_loader_test.zip (run and press end key)

**Comment by clayjohn** [comment]
> @clayjohn documented in #86384 mrp: https://github.com/godotengine/godot/files/13738417/MeshTangents.zip
> 
> and mrp from 85960: https://github.com/godotengine/godot/files/13623984/resource_loader_test.zip (run and press end key)

Thanks! I added it to the list of release blockers. I'll take a closer look during the week

**Pull request closed without merging**